File: unicodegrammars.py
from charcoaltoken import CharcoalToken as CT
import re

UnicodeGrammars = {
    CT.Arrow: [
        ["←"], ["↑"], ["→"], ["↓"], ["↖"], ["↗"], ["↘"], ["↙"]
    ],
    CT.Multidirectional: [
        [CT.Arrows, CT.Multidirectional],
        ["+", CT.Multidirectional],
        ["X", CT.Multidirectional],
        ["*", CT.Multidirectional],
        ["|", CT.Multidirectional],
        ["-", CT.Multidirectional],
        ["\\", CT.Multidirectional],
        ["/", CT.Multidirectional],
        ["<", CT.Multidirectional],
        [">", CT.Multidirectional],
        ["^", CT.Multidirectional],
        ["K", CT.Multidirectional],
        ["L", CT.Multidirectional],
        ["T", CT.Multidirectional],
        ["V", CT.Multidirectional],
        ["Y", CT.Multidirectional],
        ["7", CT.Multidirectional],
        ["¬", CT.Multidirectional],
        ["⟦", CT.Multidirectional, "⟧"],
        [CT.Separator]
    ],
    CT.Side: [
        [CT.Arrow, CT.Expression]
    ],
    CT.Separator: [
        ["¦"], []
    ],
    CT.Span: [
        [CT.Expression, "；", CT.Expression, "；", CT.Expression],
        [CT.Expression, "；", "；", CT.Expression],
        [CT.Expression, "；", CT.Expression],
        [CT.Expression, "；"],
        ["；", CT.Expression, "；", CT.Expression],
        ["；", CT.Expression],
        ["；", "；", CT.Expression],
        ["；", "；"]
    ],

    CT.Arrows: [
        [CT.Arrow, CT.Arrows],[CT.Arrow]
    ],
    CT.Sides: [
        [CT.Side, CT.Sides], [CT.Side]
    ],
    CT.Expressions: [
        [CT.Expression, CT.Expressions],
        [CT.Expression]
    ],
    CT.WolframExpressions: [
        [CT.WolframExpression, CT.WolframExpressions],
        [CT.WolframExpression]
    ],
    CT.PairExpressions: [
        [CT.Expression, CT.Expression, CT.PairExpressions],
        [CT.Expression, CT.Expression]
    ],
    CT.Cases: [
        [CT.Expression, CT.Body, CT.Cases], []
    ],

    CT.WolframList: [
        ["⟦", CT.WolframExpressions, "⟧"], ["⟦", "⟧"]
    ],
    CT.List: [
        ["⟦", CT.Expressions, "⟧"], ["⟦", "⟧"]
    ],
    CT.Dictionary: [
        ["⦃", CT.PairExpressions, "⦄"], ["⦃", "⦄"]
    ],

    CT.WolframExpression: [
        [CT.Span, CT.Separator],
        [CT.Expression]
    ],
    CT.Expression: [
        [CT.Number, CT.Separator],
        [CT.String, CT.Separator],
        [CT.Name, CT.Separator],
        [CT.List, CT.Separator],
        [CT.Dictionary, CT.Separator],
        ["«", CT.Program, "»", CT.Separator],
        [CT.OtherOperator, CT.Separator],
        [
            CT.LazyQuarternary, CT.Expression, CT.Expression, CT.Expression,
            CT.Expression
        ],
        [
            CT.Quarternary, CT.Expression, CT.Expression, CT.Expression,
            CT.Expression
        ],
        [CT.LazyTernary, CT.Expression, CT.Expression, CT.Expression],
        [CT.Ternary, CT.Expression, CT.Expression, CT.Expression],
        [CT.LazyBinary, CT.Expression, CT.Expression],
        [CT.Binary, CT.Expression, CT.Expression],
        [CT.LazyUnary, CT.Expression],
        [CT.Unary, CT.Expression],
        [CT.Nilary, CT.Separator]
    ],
    CT.Nilary: [
        ["Ｓ"], ["Ｎ"], ["‽"], ["ＫＡ"], ["ＫＭ"], ["ＫＶ"], ["ＫＫ"]
    ],
    CT.Unary: [
        ["±"], ["Ｌ"], ["¬"], ["Ｉ"], ["‽"], ["Ｖ"], ["⊟"], ["↧"], ["↥"], ["⌊"],
        ["⌈"], ["℅"], ["⮌"], ["≕"], ["″"], ["‴"], ["✂"], ["…·"], ["…"], ["～"]
    ],
    CT.Binary: [
        ["⁺"], ["⁻"], ["×"], ["÷"], ["∕"], ["﹪"], ["⁼"], ["‹"], ["›"], ["＆"],
        ["｜"], ["…·"], ["…"], ["Ｘ"], ["§"], ["⊞Ｏ"], ["⪫"], ["⪪"], ["⌕Ａ"],
        ["⌕"], ["◧"], ["◨"], ["№"], ["➙"], ["⧴"], ["？"], ["✂"], ["⊙"],
        ["⬤"]
    ],
    CT.Ternary: [
        ["✂"]
    ],
    CT.Quarternary: [
        ["✂"]
    ],
    CT.LazyUnary: [
    ],
    CT.LazyBinary: [
        ["∧"], ["∨"]
    ],
    CT.LazyTernary: [
        ["⎇"]
    ],
    CT.LazyQuarternary: [
    ],
    CT.OtherOperator: [
        ["ＫＤ", CT.Expression, CT.Arrow],
        ["Ｅ", CT.Expression, CT.Expression],
        ["▷", CT.Expression, CT.WolframList],
        ["▷", CT.Expression, CT.WolframExpression]
    ],

    CT.Program: [
        [CT.Command, CT.Separator, CT.Program],
        []
    ],
    CT.Body: [
        ["«", CT.Program, "»"],
        [CT.Command, CT.Separator]
    ],
    CT.Command: [
        ["Ｓ", CT.Name],
        ["Ｎ", CT.Name],
        ["Ｖ", CT.Expression],
        [CT.Arrow, CT.Expression],
        [CT.Expression],
        ["Ｐ", CT.Multidirectional, CT.Expression],
        ["Ｐ", CT.Expression],
        ["Ｇ", CT.Sides, CT.Expression],
        ["Ｇ", CT.Multidirectional, CT.Expression, CT.Expression],
        ["ＧＨ", CT.Sides, CT.Expression],
        ["ＧＨ", CT.Multidirectional, CT.Expression, CT.Expression],
        ["ＵＲ", CT.Expression, CT.Expression],
        ["ＵＲ", CT.Expression],
        ["ＵＯ", CT.Expression, CT.Expression, CT.Expression],
        ["ＵＯ", CT.Expression, CT.Expression],
        ["Ｂ", CT.Expression, CT.Expression, CT.Expression],
        ["Ｂ", CT.Expression, CT.Expression],
        [CT.Arrow],
        ["Ｍ", CT.Arrow],
        ["Ｍ", CT.Expression, CT.Arrow],
        ["Ｍ", CT.Expression, CT.Expression],
        ["↶", CT.Expression],
        ["↶"],
        ["↷", CT.Expression],
        ["↷"],
        ["Ｊ", CT.Expression, CT.Expression],
        ["⟲Ｔ", CT.Expression],
        ["⟲Ｔ"],
        ["‖Ｔ", CT.Multidirectional],
        ["‖Ｔ", CT.Arrow],
        ["‖Ｔ"],
        ["⟲Ｐ", CT.Arrow, CT.Number, CT.Separator],
        ["⟲Ｐ", CT.Arrow, CT.Expression],
        ["⟲Ｐ", CT.Arrow],
        ["⟲Ｐ", CT.Number, CT.Separator],
        ["⟲Ｐ", CT.Expression],
        ["⟲Ｐ"],
        ["‖Ｍ", CT.Multidirectional],
        ["‖Ｍ", CT.Arrow],
        ["‖Ｍ"],
        ["⟲Ｃ", CT.Arrow, CT.Number, CT.Separator],
        ["⟲Ｃ", CT.Arrow, CT.Expression],
        ["⟲Ｃ", CT.Arrow],
        ["⟲Ｃ", CT.Number, CT.Separator],
        ["⟲Ｃ", CT.Expression],
        ["⟲Ｃ"],
        ["‖Ｃ", CT.Multidirectional],
        ["‖Ｃ", CT.Arrow],
        ["‖Ｃ"],
        ["⟲ＯＯ", CT.Arrow, CT.Number, CT.Separator, CT.Expression],
        ["⟲ＯＯ", CT.Arrow, CT.Expression, CT.Expression],
        ["⟲ＯＯ", CT.Arrow, CT.Expression],
        ["⟲ＯＯ", CT.Number, CT.Separator, CT.Expression],
        ["⟲ＯＯ", CT.Expression, CT.Expression],
        ["⟲ＯＯ", CT.Expression],
        ["⟲Ｏ", CT.Arrow, CT.Number, CT.Separator ],
        ["⟲Ｏ", CT.Arrow, CT.Expression],
        ["⟲Ｏ", CT.Arrow],
        ["⟲Ｏ", CT.Number, CT.Separator],
        ["⟲Ｏ", CT.Expression],
        ["⟲Ｏ"],
        ["⟲ＳＯ", CT.Arrow, CT.Number, CT.Separator, CT.Expression],
        ["⟲ＳＯ", CT.Arrow, CT.Expression, CT.Expression],
        ["⟲ＳＯ", CT.Arrow, CT.Expression],
        ["⟲ＳＯ", CT.Number, CT.Separator, CT.Expression],
        ["⟲ＳＯ", CT.Expression, CT.Expression],
        ["⟲ＳＯ", CT.Expression],
        ["⟲Ｓ", CT.Arrow, CT.Number, CT.Separator],
        ["⟲Ｓ", CT.Arrow, CT.Expression],
        ["⟲Ｓ", CT.Arrow],
        ["⟲Ｓ", CT.Number, CT.Separator],
        ["⟲Ｓ", CT.Expression],
        ["⟲Ｓ"],
        ["‖ＯＯ", CT.Multidirectional, CT.Expression],
        ["‖ＯＯ", CT.Arrow, CT.Expression],
        ["‖ＯＯ", CT.Expression],
        ["‖Ｏ", CT.Multidirectional],
        ["‖Ｏ", CT.Arrow],
        ["‖Ｏ"],
        ["‖ＢＯ", CT.Multidirectional, CT.Expression],
        ["‖ＢＯ", CT.Arrow, CT.Expression],
        ["‖ＢＯ", CT.Expression],
        ["‖Ｂ", CT.Multidirectional],
        ["‖Ｂ", CT.Arrow],
        ["‖Ｂ"],
        ["⟲", CT.Expression],
        ["⟲"],
        ["‖", CT.Arrow],
        ["‖"],
        ["Ｃ", CT.Expression, CT.Expression],
        ["Ｆ", CT.Expression, CT.Body],
        ["Ｗ", CT.Expression, CT.Body],
        ["¿", CT.Expression, CT.Body, CT.Body],
        ["¿", CT.Expression, CT.Body],
        ["Ａ§", CT.Expression, CT.Expression, CT.Expression],
        ["Ａ", CT.Expression, CT.Name],
        ["Ａ", CT.Expression, CT.Expression],
        ["¤", CT.Expression],
        ["ＵＢ", CT.Expression],
        ["Ｄ"],
        ["ＲＦ", CT.Expression, CT.Expression, CT.Body],
        ["ＲＷ", CT.Expression, CT.Expression, CT.Body],
        ["Ｒ", CT.Expression],
        ["Ｒ"],
        ["ＵＴ"],
        ["Ｔ", CT.Expression, CT.Expression],
        ["Ｔ", CT.Expression],
        ["⎚"],
        ["ＵＥ", CT.Expression, CT.Expression],
        ["ＵＥ", CT.Expression],
        ["⊞", CT.Expression, CT.Expression],
        ["≡", CT.Expression, CT.Cases, CT.Body],
        ["≡", CT.Expression, CT.Cases],
        ["ＵＭ", CT.Expression, CT.Expression],
        ["▶", CT.Expression, CT.WolframList],
        ["▶", CT.Expression, CT.WolframExpression],
        ["≔", CT.Expression, CT.Expression]
    ]
}

File: js/unicodegrammars.js
'use strict';

var UnicodeGrammars = new Array(CharcoalToken.Maximum + 1);

UnicodeGrammars[CharcoalToken.Arrow] = [
    ['←'],
    ['↑'],
    ['→'],
    ['↓'],
    ['↖'],
    ['↗'],
    ['↘'],
    ['↙']
];

UnicodeGrammars[CharcoalToken.Multidirectional] = [
    [CharcoalToken.Arrows],
    ['+', CharcoalToken.Multidirectional],
    ['X', CharcoalToken.Multidirectional],
    ['*', CharcoalToken.Multidirectional],
    ['|', CharcoalToken.Multidirectional],
    ['-', CharcoalToken.Multidirectional],
    ['\\', CharcoalToken.Multidirectional],
    ['/', CharcoalToken.Multidirectional],
    ['<', CharcoalToken.Multidirectional],
    ['>', CharcoalToken.Multidirectional],
    ['^', CharcoalToken.Multidirectional],
    ['K', CharcoalToken.Multidirectional],
    ['L', CharcoalToken.Multidirectional],
    ['T', CharcoalToken.Multidirectional],
    ['V', CharcoalToken.Multidirectional],
    ['Y', CharcoalToken.Multidirectional],
    ['7', CharcoalToken.Multidirectional],
    ['¬', CharcoalToken.Multidirectional],
    []
];

UnicodeGrammars[CharcoalToken.Side] = [
    [CharcoalToken.Arrow, CharcoalToken.Expression]
];

UnicodeGrammars[CharcoalToken.Separator] = [
    ['¦'],
    []
];

UnicodeGrammars[CharcoalToken.Arrows] = [
    [CharcoalToken.Arrow, CharcoalToken.Arrows],
    [CharcoalToken.Arrow]
];

UnicodeGrammars[CharcoalToken.Sides] = [
    [CharcoalToken.Side, CharcoalToken.Sides],
    [CharcoalToken.Side]
];

UnicodeGrammars[CharcoalToken.Expressions] = [
    [
        CharcoalToken.Expression,
        CharcoalToken.Expressions
    ],
    [CharcoalToken.Expression]
];

UnicodeGrammars[CharcoalToken.PairExpressions] = [
    [
        CharcoalToken.Expression,
        CharcoalToken.Expression,
        CharcoalToken.PairExpressions
    ],
    [CharcoalToken.Expression, CharcoalToken.Expression]
];

UnicodeGrammars[CharcoalToken.List] = [
    ['⟦', CharcoalToken.Expressions, '⟧'],
    ['⟦', '⟧']
];

UnicodeGrammars[CharcoalToken.ArrowList] = [
    ['⟦', CharcoalToken.Multidirectional, '⟧'],
    ['⟦', '⟧']
];

UnicodeGrammars[CharcoalToken.Dictionary] = [
    ['⦃', CharcoalToken.PairExpressions, '⦄'],
    ['⦃', '⦄']
];

UnicodeGrammars[CharcoalToken.Expression] = [
    [CharcoalToken.Number, CharcoalToken.Separator],
    [CharcoalToken.String, CharcoalToken.Separator],
    [CharcoalToken.Name, CharcoalToken.Separator],
    [CharcoalToken.List, CharcoalToken.Separator],
    [CharcoalToken.Dictionary, CharcoalToken.Separator],
    [
        CharcoalToken.OtherOperator,
        CharcoalToken.Separator
    ],
    [
        CharcoalToken.LazyTernary,
        CharcoalToken.Expression,
        CharcoalToken.Expression,
        CharcoalToken.Expression
    ],
    [
        CharcoalToken.Ternary,
        CharcoalToken.Expression,
        CharcoalToken.Expression,
        CharcoalToken.Expression
    ],
    [
        CharcoalToken.LazyBinary,
        CharcoalToken.Expression,
        CharcoalToken.Expression
    ],
    [
        CharcoalToken.Binary,
        CharcoalToken.Expression,
        CharcoalToken.Expression
    ],
    [
        CharcoalToken.LazyUnary,
        CharcoalToken.Expression
    ],
    [
        CharcoalToken.Unary,
        CharcoalToken.Expression
    ],
    [CharcoalToken.Nilary, CharcoalToken.Separator]
];

UnicodeGrammars[CharcoalToken.Nilary] = [
    ['Ｓ'],
    ['Ｎ'],
    ['‽'],
    ['ＫＡ'],
    ['Ｋ']
];

UnicodeGrammars[CharcoalToken.Unary] = [
    ['⁻'],
    ['Ｌ'],
    ['¬'],
    ['Ｉ'],
    ['‽'],
    ['Ｖ'],
    ['⊟'],
    ['↧'],
    ['↥'],
    ['⌊'],
    ['⌈']
];

UnicodeGrammars[CharcoalToken.Binary] = [
    ['⁺'],
    ['⁻'],
    ['×'],
    ['÷'],
    ['﹪'],
    ['⁼'],
    ['‹'],
    ['›'],
    ['…·'],
    ['…'],
    ['Ｘ'],
    ['§'],
    ['⊞'],
    ['⪫'],
    ['⪪'],
    ['⌕Ａ'],
    ['⌕']
];

UnicodeGrammars[CharcoalToken.Ternary] = [
];

UnicodeGrammars[CharcoalToken.LazyUnary] = [
];

UnicodeGrammars[CharcoalToken.LazyBinary] = [
    ['∧'],
    ['∨']
];

UnicodeGrammars[CharcoalToken.LazyTernary] = [
    ['⎇']
];

UnicodeGrammars[CharcoalToken.OtherOperator] = [
    ['ＫＤ', CharcoalToken.Expression, CharcoalToken.Arrow]
];

UnicodeGrammars[CharcoalToken.Program] = [
    [CharcoalToken.Command, CharcoalToken.Program],
    []
];

UnicodeGrammars[CharcoalToken.Body] = [
    ['«', CharcoalToken.Program, '»'],
    [CharcoalToken.Command]
];

UnicodeGrammars[CharcoalToken.Command] = [
    ['Ｓ', CharcoalToken.Name],
    ['Ｎ', CharcoalToken.Name],
    ['Ｖ', CharcoalToken.Expression],
    [CharcoalToken.Arrow, CharcoalToken.Expression],
    [CharcoalToken.Expression],
    ['Ｐ', CharcoalToken.Multidirectional, CharcoalToken.Expression],
    ['Ｐ', CharcoalToken.Expression],
    ['Ｇ', CharcoalToken.Sides, CharcoalToken.Expression],
    [
        'Ｇ',
        CharcoalToken.Multidirectional,
        CharcoalToken.Expression,
        CharcoalToken.Expression
    ],
    ['ＧＨ', CharcoalToken.Sides, CharcoalToken.Expression],
    [
        'ＧＨ',
        CharcoalToken.Multidirectional,
        CharcoalToken.Expression,
        CharcoalToken.Expression
    ],
    ['ＢＲ', CharcoalToken.Expression, CharcoalToken.Expression],
    [
        'Ｂ',
        CharcoalToken.Expression,
        CharcoalToken.Expression,
        CharcoalToken.Expression
    ],
    [CharcoalToken.Arrow],
    ['Ｍ', CharcoalToken.Arrow],
    ['Ｍ', CharcoalToken.Expression, CharcoalToken.Arrow],
    ['↶', CharcoalToken.Expression],
    ['↶'],
    ['↷', CharcoalToken.Expression],
    ['↷'],
    ['Ｊ', CharcoalToken.Expression, CharcoalToken.Expression],
    ['⟲Ｔ', CharcoalToken.Expression],
    ['‖Ｔ', CharcoalToken.ArrowList],
    ['‖Ｔ', CharcoalToken.Arrow],
    ['‖Ｍ', CharcoalToken.ArrowList],
    ['‖Ｍ', CharcoalToken.Arrow],
    ['⟲Ｃ', CharcoalToken.Expression, CharcoalToken.Arrow],
    ['⟲Ｃ', CharcoalToken.Expression],
    ['‖Ｃ', CharcoalToken.ArrowList],
    ['‖Ｃ', CharcoalToken.Arrow],
    ['⟲Ｏ', CharcoalToken.Expression],
    ['‖Ｏ', CharcoalToken.ArrowList],
    ['‖Ｏ', CharcoalToken.Arrow],
    ['⟲', CharcoalToken.Expression],
    ['‖', CharcoalToken.Arrow],
    ['Ｃ', CharcoalToken.Expression, CharcoalToken.Expression],
    ['Ｆ', CharcoalToken.Expression, CharcoalToken.Body],
    ['Ｗ', CharcoalToken.Expression, CharcoalToken.Body],
    [
        '¿',
        CharcoalToken.Expression,
        CharcoalToken.Body,
        CharcoalToken.Body
    ],
    ['¿', CharcoalToken.Expression, CharcoalToken.Body],
    ['Ａ', CharcoalToken.Expression, CharcoalToken.Name],
    ['¤', CharcoalToken.Expression],
    ['ＵＢ', CharcoalToken.Expression],
    ['Ｄ'],
    [
        'ＨＦ',
        CharcoalToken.Expression,
        CharcoalToken.Expression,
        CharcoalToken.Body
    ],
    [
        'ＨＷ',
        CharcoalToken.Expression,
        CharcoalToken.Expression,
        CharcoalToken.Body
    ],
    ['Ｈ', CharcoalToken.Expression],
    ['Ｈ'],
    ['Ｔ', CharcoalToken.Expression, CharcoalToken.Expression],
    ['⎚'],
    ['ＵＥ', CharcoalToken.Expression, CharcoalToken.Expression],
    ['ＵＥ', CharcoalToken.Expression]
];

